libsnark/zk_proof_systems/plonk/prover.tcc

```diff
@@ -0,0 +1,989 @@
+    polynomial<Field> S_1_zeta_poly_neg{-round_four_out.S_1_zeta};
+    //    libfqfft::_polynomial_multiplication<Field>(S_1_zeta_poly_neg,
+    //    S_1_zeta_poly, neg_one_poly);
+    // (S1(x) - S_1_zeta)
+    polynomial<Field> S1_sub_szeta;
+    libfqfft::_polynomial_addition<Field>(
+        S1_sub_szeta, srs.S_polys[1], S_1_zeta_poly_neg);
+    // (S1(x) - S_1_zeta) * nu^6
+    Field nu6 = libff::power(nu, libff::bigint<1>(6));
+    polynomial<Field> nu6_poly{nu6};
+    libfqfft::_polynomial_multiplication<Field>(
+        W_zeta_part[6], S1_sub_szeta, nu6_poly);
+
+    // compute full zeta polynomial W_zeta = \sum W_zeta_part[i]
+    int nzeta = 7;
+    polynomial<Field> W_zeta(round_zero_out.null_poly);
+    for (int i = 0; i < nzeta; ++i) {
+        libfqfft::_polynomial_addition<Field>(W_zeta, W_zeta, W_zeta_part[i]);
+    }
+
+    // compute 1/(X-zeta) * W_zeta
+    polynomial<Field> x_sub_zeta_poly{-zeta, Field(1)};
+    libfqfft::_polynomial_division(W_zeta, remainder, W_zeta, x_sub_zeta_poly);
+    assert(libfqfft::_is_zero(remainder));
+
+    // Compute opening proof:
+    // W_zeta_omega = z(X) - z(zeta*srs.omega_roots) / X -
+    // (zeta*srs.omega_roots)
+    polynomial<Field> W_zeta_omega{round_zero_out.null_poly};
+
+    // -z(zeta*srs.omega_roots)
+    polynomial<Field> z_poly_xomega_zeta_neg{
+        -round_four_out.z_poly_xomega_zeta};
+    // z(X) - z(zeta*srs.omega_roots)
+    libfqfft::_polynomial_addition<Field>(
+        W_zeta_omega, round_two_out.z_poly, z_poly_xomega_zeta_neg);
+    // -zeta*srs.omega_roots; srs.omega_roots[base][1] =
+    // srs.omega_roots_base
+    polynomial<Field> x_sub_zeta_omega_roots{
+        -(zeta * srs.omega_roots[base][1]), Field(1)};
+
+    // z(X) - z(zeta*srs.omega_roots) / X -
+    // (zeta*srs.omega_roots)
+    libfqfft::_polynomial_division(
+        W_zeta_omega, remainder, W_zeta_omega, x_sub_zeta_omega_roots);
+    assert(libfqfft::_is_zero(remainder));
+
+    // TODO: make a separate unit test for W_zeta, W_zeta_omega
+#if 0
+    assert(W_zeta == example.W_zeta);
+    assert(W_zeta_omega == example.W_zeta_omega);
+#endif // #ifdef DEBUG_PLONK
+
+    // Evaluate polynomials W_zeta and W_zeta_omega at the seceret
+    // input
+    W_zeta_at_secret =
+        plonk_evaluate_poly_at_secret_G1<ppT>(srs.secret_powers_g1, W_zeta);
+    W_zeta_omega_at_secret = plonk_evaluate_poly_at_secret_G1<ppT>(
+        srs.secret_powers_g1, W_zeta_omega);
+
+    round_five_out_t<ppT> round_five_out(
+        std::move(r_zeta),
+        std::move(W_zeta_at_secret),
+        std::move(W_zeta_omega_at_secret));
+
+    return round_five_out;
+}
+
+/// Prover compute SNARK proof
+///
+/// Pi ([a]_1, [b]_1, [c]_1, [z]_1,
+///     [t_lo]_1, [t_mi]_1, [t_hi]_1,
+///     \bar{a}, \bar{b}, \bar{c},
+///     \bar{S_sigma1}, \bar{S_sigma2}, \bar{z_w},
+///     [W_zeta]_1, [W_{zeta omega}]_1
+///     r_zeta)
+///
+/// \note in the reference Python implementation, r_zeta (the
```

Comment:
Is the "reference Python implementation" the code from the blog you saw?  Or is there an actual reference impl?  We might want to just remove references to it (when all checks are moved to the unit tests).

Reply:
Understood. Thanks.